app/org/catalyst/slackservice/controllers/EventController.java

```diff
@@ -248,4 +215,117 @@ private static boolean isInvalidUrlVerification(final Request request) {
+                return ResultHelper.badRequest(messages, MessageHandler.REQUEST_NOT_VERIFIED);
+            }
+
+            @Override
+            public boolean matches(Request request, Http.Request httpRequest, AppConfig config) {
+                return !RequestVerifier.verified(httpRequest, config.getSigningSecret(), config.getToken(), request.token);
+            }
+        },
+        UrlVerification {
+            @Override
+            public CompletionStage<Result> handle(EventController controller, Http.Request httpRequest, MessageHandler messages, Request request, Logger logger) {
+                return controller.handleURLVerification(request.challenge);
+            }
+
+            @Override
+            public boolean matches(Request request, Http.Request httpRequest, AppConfig config) {
+                return request.type.equals(TYPE_URL_VERIFICATION);
+            }
+        },
+        EventCallback {
+            @Override
+            public CompletionStage<Result> handle(EventController controller, Http.Request httpRequest, MessageHandler messages, Request request, Logger logger) {
+                return controller.handleEventCallback(request.event);
+            }
+
+            @Override
+            public boolean matches(Request request, Http.Request httpRequest, AppConfig config) {
+                return request.type.equals(TYPE_EVENT_CALLBACK) && !SUBTYPE_TOKENS_REVOKED.equals(request.event.type);
+            }
+        },
+        SubtypeTokensRevoked {
+            @Override
+            public CompletionStage<Result> handle(EventController controller, Http.Request httpRequest, MessageHandler messages, Request request, Logger logger) {
+                return controller.handleRevokeTokens(request.teamId, request.event);
+            }
+
+            @Override
+            public boolean matches(Request request, Http.Request httpRequest, AppConfig config) {
+                return request.type.equals(TYPE_EVENT_CALLBACK) && SUBTYPE_TOKENS_REVOKED.equals(request.event.type);
+            }
+        },
+        UnsupportedType {
+            @Override
+            public CompletionStage<Result> handle(EventController controller, Http.Request httpRequest, MessageHandler messages, Request request, Logger logger) {
+                return ResultHelper.badRequest(messages, MessageHandler.UNSUPPORTED_TYPE);
+            }
+
+            @Override
+            public boolean matches(Request request, Http.Request httpRequest, AppConfig config) {
+                return false; /* Fallback type, should never match */
+            }
+        };
+
+        public abstract CompletionStage<Result> handle(EventController controller, Http.Request httpRequest, MessageHandler messages, Request request, Logger logger);
+        public abstract boolean matches(Request request, Http.Request httpRequest, AppConfig config);
+
+        public static RequestHandler GetHandler(Request request, Http.Request httpRequest, AppConfig config) {
```

Comment:
I realize now that my C# background is showing.  Capital G in `GetHandler`?  Yeah, those are Microsoft's recommended naming conventions.